File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 2/구명보트/구명보트.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> people, int limit) {
    // 사람 몸무게를 내림차순으로 정렬
    sort(people.begin(), people.end(), greater<int>());
    
    // 투포인터 이용
    // 가장 무거운 사람 + 가장 가벼운 사람 조합
    int cnt=0, left=0, right=people.size()-1;
    while(left < right){
        // 둘 다 태우기
        if(people[left]+people[right] <= limit){
            left++; right--;
            cnt++;
        }
        // 아니면 무거운 사람부터 태우기
        else{
            left++; cnt++;
        }
    }
    
    // 마지막 사람 남았으면 태우기
    if(left == right) cnt++;
    
    return cnt;
}

File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 2/구명보트/구명보트.py
def solution(people, limit):
    people = sorted(people, reverse=True)

    cnt,left = 0,0; right=len(people)-1
    while left < right:
        if people[left]+people[right] <= limit:
            left += 1; right -= 1
            cnt += 1
        else:
            left += 1
            cnt += 1

    if left == right: cnt += 1

    return cnt
